tests/simple/testdata/type_deduction.textproto

```diff
@@ -259,3 +259,497 @@ section {
+        abstract_type {
+          name: "tuple"
+          parameter_types { dyn {} }
+          parameter_types { dyn {} }
+          parameter_types { dyn {} }
+        }
+      }
+    }
+  }
+}
+
+# Messages, some well-known types, and abstract can be assigned from null in the
+# CEL type checkers for historical reasons. For backward compatibiliy, we
+# maintain this behavior but it's a bit inconsistent with the intended meaning
+# of null (a singleton representing a JSON null literal) in CEL.
+section {
+  name: "legacy_nullable_types"
+  test {
+    name: "null_assignable_to_message_parameter_candidate"
+    type_env {
+      name: "msg"
+      ident {
+        type {
+          message_type: "cel.expr.conformance.proto3.TestAllTypes"
+        }
+      }
+    }
+    bindings {
+      key: "msg"
+      value {
+        value {
+          object_value {
+            [type.googleapis.com/cel.expr.conformance.proto3.TestAllTypes] {}
+          }
+        }
+      }
+    }
+    expr: "[msg, null][0]"
+    typed_result {
+      deduced_type {
+        message_type: "cel.expr.conformance.proto3.TestAllTypes"
+      }
+      result {
+        object_value {
+          [type.googleapis.com/cel.expr.conformance.proto3.TestAllTypes] {}
+        }
+      }
+    }
+  }
+  test {
+    name: "null_assignable_to_duration_parameter_candidate"
+    type_env {
+      name: "msg"
+      ident {
+        type {
+          message_type: "cel.expr.conformance.proto3.TestAllTypes"
+        }
+      }
+    }
+    bindings {
+      key: "msg"
+      value {
+        value {
+          object_value {
+            [type.googleapis.com/cel.expr.conformance.proto3.TestAllTypes] {}
+          }
+        }
+      }
+    }
+    expr: "[msg.single_duration, null][0]"
+    typed_result {
+      deduced_type {
+        well_known: DURATION
+      }
+      result {
+        object_value {
+          [type.googleapis.com/google.protobuf.Duration] {}
+        }
+      }
+    }
+  }
+  test {
+    name: "null_assignable_to_timestamp_parameter_candidate"
+    type_env {
+      name: "msg"
+      ident {
+        type {
+          message_type: "cel.expr.conformance.proto3.TestAllTypes"
+        }
+      }
+    }
+    bindings {
+      key: "msg"
+      value {
+        value {
+          object_value {
+            [type.googleapis.com/cel.expr.conformance.proto3.TestAllTypes] {}
+          }
+        }
+      }
+    }
+    expr: "[msg.single_timestamp, null][0]"
+    typed_result {
+      deduced_type {
+        well_known: TIMESTAMP
+      }
+      result {
+        object_value {
+          [type.googleapis.com/google.protobuf.Timestamp] {}
+        }
+      }
+    }
+  }
+  test {
+    name: "null_assignable_to_abstract_parameter_candidate"
+    expr: "[optional.of(1), null][0]"
+    check_only: true
+    typed_result {
+      deduced_type {
+        abstract_type {
+          name: "optional_type"
+          parameter_types { primitive: INT64 }
+        }
+      }
+    }
```

Comment:
Null is a bit weird, it's meant to be a monotype for a singleton so I'd argue that it's more specific. The legacy nullability rules seem to behave more as though the message-like types are implicitly (null | Type) instead of null being a more general type. I think the go implementation doesn't consider null as strictly more general so it is order dependent on which type it will report for the list. I'm open to adjusting, but not sure what's right here since this is an existing but not quite consistent part of null handling

Reply:
Interesting, thanks for elaborating -- I'll defer to @TristonianJones on what the intended behavior is.

Reply:
@jnthntatum is right, the handling of `null` precedence and general `null | Type` behavior is a wart in how CEL was originally implemented and then propagated. Ideally, we could address `null` handling and precedence as tech debt to resolve next year.